File: 110.balanced-binary-tree.js
/*
 * @lc app=leetcode id=110 lang=javascript
 *
 * [110] Balanced Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
  if (!root) return true;
  return height(root);

  function height(node) {
    if (!node) return 0;
    let leftHeight = height(node.left);
    let rightHeight = height(node.right);
    if (leftHeight === false || rightHeight === false) {
      return false;
    }

    if (Math.abs(leftHeight - rightHeight) > 1) {
      return false;
    }
    return Math.max(leftHeight, rightHeight) + 1;
  }
};
// @lc code=end
// 227/227 cases passed (64 ms)
// Your runtime beats 89 % of javascript submissions
// Your memory usage beats 85.71 % of javascript submissions (37.5 MB)

File: 110.balanced-binary-tree.java
/*
 * @lc app=leetcode id=110 lang=java
 *
 * [110] Balanced Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        if (root == null)
            return true;
        return height(root) != -1;
    }

    public int height(TreeNode node) {
        if (node == null)
            return 0;
        int leftHeight = height(node.left);
        int rightHeight = height(node.right);
        if (leftHeight == -1 || rightHeight == -1) {
            return -1;
        }
        if (Math.abs(leftHeight - rightHeight) > 1) {
            return -1;
        }
        return Math.max(leftHeight, rightHeight) + 1;
    }

}
// @lc code=end
// 227/227 cases passed (1 ms)
// Your runtime beats 98.66 % of java submissions
// Your memory usage beats 87.96 % of java submissions (38.6 MB)